### PR TITLE
Push Picasso down into RequestHandler

### PR DESCRIPTION
**Pull request by jrodbx**
Trying to figure out why Mockito tests are failing...

### CODE REVIEW

File: picasso/src/main/java/com/squareup/picasso3/RequestHandler.java [hunk 1]
```diff
@@ -107,11 +108,12 @@ int getExifOrientation() {
 
   /**
    * Loads an image for the given {@link Request}.
+   * @param picasso
```
Comment (swankjesse) [inline picasso/src/main/java/com/squareup/picasso3/RequestHandler.java]: nit: unnecessary

File: picasso/src/main/java/com/squareup/picasso3/RequestHandler.java [hunk 2]
```diff
@@ -107,11 +108,12 @@ int getExifOrientation() {
...
    * @param request the data from which the image should be resolved.
    * @param networkPolicy the {@link NetworkPolicy} for this request.
    * @param callback
```
Comment (swankjesse) [inline picasso/src/main/java/com/squareup/picasso3/RequestHandler.java]: same

**Comment by swankjesse** [comment]
Does `eq()` work if the argument to it is itself a mock? Hmm

File: picasso/src/main/java/com/squareup/picasso3/RequestHandler.java (hunk 2 above)
Comment (jrodbx) [reply picasso/src/main/java/com/squareup/picasso3/RequestHandler.java]: updated.

File: picasso/src/main/java/com/squareup/picasso3/RequestHandler.java (hunk 1 above)
Comment (jrodbx) [reply picasso/src/main/java/com/squareup/picasso3/RequestHandler.java]: updated.

**Comment by jrodbx** [comment]
Good catch; however, changing `picasso` from a mock to a real impl doesn't resolve the issue.

**Comment by runningcode** [comment]
Ok, no idea why this works, but it does
```
diff --git a/picasso/src/test/java/com/squareup/picasso3/BitmapHunterTest.java b/picasso/src/test/java/com/squareup/picasso3/BitmapHunterTest.java
index 6f99a67..90277f2 100644
--- a/picasso/src/test/java/com/squareup/picasso3/BitmapHunterTest.java
+++ b/picasso/src/test/java/com/squareup/picasso3/BitmapHunterTest.java
@@ -161,9 +161,10 @@ public final class BitmapHunterTest {
         new TestableBitmapHunter(picasso, dispatcher, cache, stats, action, bitmap);

     Bitmap result = hunter.hunt();
+    Request request = action.getRequest();
     assertThat(cache.missCount()).isEqualTo(1);
     verify(hunter.requestHandler).load(
-        eq(picasso), eq(action.getRequest()), eq(0), any(RequestHandler.Callback.class));
+        eq(picasso), eq(request), eq(0), any(RequestHandler.Callback.class));
     assertThat(result).isEqualTo(bitmap);
   }

@@ -179,8 +180,9 @@ public final class BitmapHunterTest {
         new TestableBitmapHunter(picasso, dispatcher, cache, stats, action, bitmap);

     Bitmap result = hunter.hunt();
+    Request request = action.getRequest();
     assertThat(cache.hitCount()).isEqualTo(1);
-    verify(hunter.requestHandler, never()).load(eq(picasso), eq(action.getRequest()), eq(0),
+    verify(hunter.requestHandler, never()).load(eq(picasso), eq(request), eq(0),
         any(RequestHandler.Callback.class));
     assertThat(result).isEqualTo(bitmap);
   }
```

**Comment by jrodbx** [comment]
Ahh, I see.  Looks Mockito doesn't like it when you interact with a mock (here, by calling `action.getRequest()`) while in the middle of verifying a mock (here, `hunter.requestHandler`) using argument matchers.  The problem didn't exist before since `action.getRequest()` used to be the first argument and therefore was the first matcher encountered during verification.

Extracting the call to a variable moves the interaction with `action` to a point prior to the verification of `hunter.requestHandler.load(...)`.  Moving the newly-added `picasso` parameter away from the first position (i.e., after `action.getRequest()`) also "fixes" the issue.

Thanks @runningcode!

**Pull request merged**